quil-rs/src/bin/stub_gen.rs

```diff
@@ -3,12 +3,316 @@
 //! see the [`pyo3-stub-gen`][] documentation.
 //!
 //! [`pyo3-stub-gen`]: https://github.com/Jij-Inc/pyo3-stub-gen
+
 #[cfg(feature = "stubs")]
-use pyo3_stub_gen::Result;
+/// Provide support for sorting `.pyi` stubs.
+///
+/// [`pyo3-stub-gen`] doesn't keep its output in a consistent order.  Thankfully the order is
+/// deterministic if the program is deterministic, but due to the use of the [`inventory`][] crate,
+/// it's not deterministic if the code changes.  This module thus sorts the components of stubs that
+/// are not already in a fixed order.
+///
+/// [`pyo3-stub-gen`]: https://github.com/Jij-Inc/pyo3-stub-gen
+mod sort_stubs {
+    use std::{
+        any::TypeId,
+        cmp::Ordering,
+        collections::{BTreeMap, BTreeSet, HashSet},
+        path::PathBuf,
+    };
+
+    use indexmap::IndexMap;
+    use itertools::Itertools as _;
+    use pyo3_stub_gen::{
+        generate::{
+            Arg, ClassDef, EnumDef, FunctionDef, MemberDef, MethodDef, MethodType, Module,
+            VariableDef,
+        },
+        type_info::{DeprecatedInfo, IgnoreTarget, SignatureArg},
+        StubInfo, TypeInfo,
+    };
+
+    /// A trait that's equivalent to [`Ord`] but not semantically meaningful, used for putting
+    /// [`pyo3_stub_gen`] types in a consistent order.
+    trait ArbitraryOrd {
+        /// Analogous to [`Ord::cmp`]
+        #[must_use]
+        fn cmp(&self, other: &Self) -> Ordering;
+    }
+
+    /// Wrap references to an [`Ord`] type as an [`ArbitraryOrd`] type.
+    struct Arbitrary<'a, T>(&'a T);
+
+    impl<T: ArbitraryOrd> PartialEq for Arbitrary<'_, T> {
+        fn eq(&self, other: &Self) -> bool {
+            self.cmp(other) == Ordering::Equal
+        }
+    }
+
+    impl<T: ArbitraryOrd> Eq for Arbitrary<'_, T> {}
+
+    impl<T: ArbitraryOrd> PartialOrd for Arbitrary<'_, T> {
+        fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
+            Some(self.cmp(other))
+        }
+    }
+
+    impl<T: ArbitraryOrd> Ord for Arbitrary<'_, T> {
+        fn cmp(&self, other: &Self) -> Ordering {
+            self.0.cmp(other.0)
+        }
+    }
+
+    macro_rules! arbitrary_ord_structs {
+        ($(
+            $struct:ident { $($field:ident),* $(,)? };
+        )*) => {
+            $(
+                #[automatically_derived]
+                impl ArbitraryOrd for $struct {
+                    fn cmp(&self, other: &Self) -> Ordering {
+                        // This guarantees us exhaustiveness
+                        let $struct { $($field),* } = self;
+
+                        let result = Ordering::Equal;
```

Comment:
Minor suggestion: put a comment here like

```suggestion
                        // Return the first non-Equal result when comparing field pairs,
                        // or return Equal if all field pairs compare Equal.
                        let result = Ordering::Equal;
```